Implement the patient demographics contract

rebkwok opened the pull request:
This PR updates the Patient table contract to reflect the real contract, with column constraints, and adds a `patient_demographics` table for the TPP backend which implements this contract.

Also adds a Constraint class, with the beginnings of a method for validating the data in backend databases based on those constraints.

rebkwok commented on cohortextractor2/backends/tpp.py:
```diff
@@ -59,14 +59,32 @@ class TPPBackend(BaseBackend):
+        ),
+        query="""
+            SELECT  Patient_ID as patient_id,
+                datefromparts(year(DateOfBirth), month(DateOfBirth), 1) as date_of_birth,
```
This may not be necessary if TPP dates are already in Y-M-01 format

rebkwok commented on cohortextractor2/concepts/table_contract.py:
```diff
@@ -1,13 +1,16 @@
 class Column:
 
     type: BaseType  # noqa: A003
+    description: str  # noqa: A003
```
I've added description, and kept the help field for notes about constraints.

rebkwok commented on cohortextractor2/concepts/tables.py:
```diff
@@ -21,30 +22,40 @@ def __init__(self):
-        help="The patient's date of birth",
+        description="Patient's year and month of birth.",
+        help="The day will always be the first of the month. Must be present.",
+        constraints=[DateConstraint(match_format=["%Y-%m-01"])],
```
Where do we enforce required fields? Should we have a` RequiredConstraint` that checks there are no nulls in a column?

benbc replied:
Yes, I think that's another kind of constraint.

benbc commented on cohortextractor2/concepts/tables.py:
```diff
@@ -21,30 +22,40 @@ def __init__(self):
+            "including but not limited to 'rather not say' and empty/missing values). "
+            "Must be present."
+        ),
+        constraints=[ChoiceConstraint()],
```
Can this be implied by the type, rather than having to duplicate it?

rebkwok replied:
yes, will do that

benbc commented on cohortextractor2/concepts/constraints.py:
```diff
@@ -0,0 +1,42 @@
+        """
+        match_format: str.  A constraint format in the form '%Y-%m-%d' to compare with the base format.
+        e.g. '%Y-%m-01' ensures that each date in the backend column repesents the 1st of a month
+        """
```
I'm doubtful that the generality you're providing for here will _really_ prove to be useful. Maybe better to have a date-is-first-of-month constraint, which might also make generating documentation for it easier, them generalize later if there are other similar constraints that we need to support.

rebkwok replied:
OK, sure, that makes sense

benbc commented on cohortextractor2/backends/tpp.py:
```diff
@@ -59,14 +59,32 @@ class TPPBackend(BaseBackend):
+                    ELSE 'unknown'
+                END as sex
+            FROM Patient
+        """,
```
It's great that this is fairly simple and that the existing `QueryTable` supports what we need.

benbc reviewed (commented):
This looks like a good start. We've got a story coming later to give backend implementers a way to run validation of their actual data, which I guess will hook into that stuff you're adding. Might be good to give a little thought to that now, as you are doing, but don't worry that it's not actually used in practice.

rebkwok commented on cohortextractor2/concepts/tables.py:
```diff
@@ -21,30 +22,40 @@ def __init__(self):
             "output or operate on this column"
         ),
+        help="",
+        constraints=[NotNullConstraint(), UniqueConstraint()],
```
@benbc since patient_id isn't necessarily the primary key in a backend table, I think we probably do want to check this?

benbc replied:
:+1:

rebkwok commented:
@benbc OK, I think I am done with this PR for now.  I'll leave it in draft atm while the patient demographics contract is still being finalised.  I've added DateOfDeath, and replaced ceiling dates with nulls. We can easily remove that if we decide not to put that in the contract, and I think there will be some help/description wording to update if we do.  If we don't I assume we'll want to have backends normalise their placeholder-for-null value to a consistent one, so there will still be some help/description wording to update.

rebkwok commented:
Making this non-draft now, since it seems there are a bunch of contract questions that will be pending for a while yet

rebkwok commented on cohortextractor2/backends/tpp.py:
```diff
@@ -59,14 +59,38 @@ class TPPBackend(BaseBackend):
+            SELECT  Patient_ID as patient_id,
+                DateOfBirth as date_of_birth,
+                CASE
+                    WHEN DateOfDeath = '9999-12-31' THEN NULL
```
This will need to be addressed again later; we may want to keep ceiling dates (in which case we need to not replace the day, and validation needs to deal with them), and/or we may need to replace nulls with ceiling dates

benbc reviewed (approved):
Looks great.

rebkwok commented on Justfile:
```diff
@@ -150,6 +150,5 @@ test-all *ARGS: devenv build-cohort-extractor
         --cov=tests \
         --cov-report=html \
         --cov-report=term-missing:skip-covered \
-        tests \
```
I've changed it now so that "tests" is the default ARGS value - so it'll do what it did before by default, but still let you run just specific test(s) if you want to.

ghickman replied:
Simon just hit the same thing in https://github.com/opensafely-core/cohort-extractor-v2/pull/278 and solved with adding this path to the config in pyproject.toml.  I think I'm very gently leaning towards the config option, only because it fits with our general ethos of default config in pyproject.toml, happy to be persuaded otherwise though!

rebkwok replied:
I don't have a strong feeling either way, as long as I can run individual tests locally :)
